core/src/shred_fetch_stage.rs

```diff
@@ -1,43 +1,103 @@
+
+        if index_end <= p.meta.size {
+            if let Ok(index) = limited_deserialize::<u32>(&p.data[index_start..index_end]) {
+                if index < MAX_DATA_SHREDS_PER_SLOT as u32 && slot_end <= p.meta.size {
+                    if let Ok(slot) = limited_deserialize::<Slot>(&p.data[slot_start..slot_end]) {
+                        return Some((slot, index));
+                    }
+                }
+            }
+        } else {
+            *index_overrun += 1;
+        }
+        None
+    }
+
     // updates packets received on a channel and sends them on another channel
-    fn modify_packets<F>(recvr: PacketReceiver, sendr: PacketSender, modify: F)
-    where
+    fn modify_packets<F>(
+        recvr: PacketReceiver,
+        sendr: PacketSender,
+        bank_forks: Option<Arc<RwLock<BankForks>>>,
+        modify: F,
+    ) where
         F: Fn(&mut Packet),
     {
+        let mut shreds_received = ShredsReceived::default();
+        let mut last_cleared = Instant::now();
+
+        // In the case of bank_forks=None, setup to accept any slot range
+        let mut last_root = 0;
+        let mut last_slot = std::u64::MAX;
+        let mut slots_per_epoch = 0;
+
         while let Some(mut p) = recvr.iter().next() {
-            let index_start = OFFSET_OF_SHRED_INDEX;
-            let index_end = index_start + SIZE_OF_SHRED_INDEX;
+            if last_cleared.elapsed().as_millis() > 200 {
+                shreds_received.clear();
+                last_cleared = Instant::now();
+                if let Some(bank_forks) = bank_forks.as_ref() {
+                    let bank_forks_r = bank_forks.read().unwrap();
+                    last_root = bank_forks_r.root();
+                    let working_bank = bank_forks_r.working_bank();
+                    last_slot = working_bank.slot();
+                    let root_bank = bank_forks_r.root_bank();
+                    slots_per_epoch = root_bank.get_slots_in_epoch(root_bank.epoch());
+                }
+            }
+            let mut index_overrun = 0;
+            let mut shred_count = 0;
             p.packets.iter_mut().for_each(|p| {
+                shred_count += 1;
                 p.meta.discard = true;
-                if index_end <= p.meta.size {
-                    if let Ok(index) = limited_deserialize::<u32>(&p.data[index_start..index_end]) {
-                        if index < MAX_DATA_SHREDS_PER_SLOT as u32 {
+                if let Some((slot, index)) = Self::get_slot_index(p, &mut index_overrun) {
+                    if slot > last_root && slot < (last_slot + 2 * slots_per_epoch) {
+                        // Shred filter
+                        let slot_received = shreds_received.entry(slot).or_insert_with(|| {
+                            BitVec::new_fill(false, MAX_DATA_SHREDS_PER_SLOT as u64)
+                        });
+                        if !slot_received.get(index.into()) {
                             p.meta.discard = false;
```

Comment:
isn't the discard flag `false` by default?

Reply:
No, it's set to true by default earlier in the loop.
https://github.com/solana-labs/solana/pull/8975/files#diff-6e3721caf0a126d8a8852e4f448fc0e7R84

Reply:
ah missed that, coooooo